main.js

```diff
@@ -1,3 +1,83 @@
 import './style.css';
 
-console.log('Hello world');
+function getRect(el) {
+    const rect = el.getBoundingClientRect();
+    return {
+        left: rect.left + window.scrollX,
+        top: rect.top + window.scrollY,
+        width: rect.width,
+        height: rect.height
+    };
+}
+
+function setTipPosition(element, tip) {
```

Comment:
At first i should say "Wow", cause you made good code with all features I have mentioned about homework and I see that you are kinda experienced developer in frontend.
But i will highlight something to let you just know :)